File: app/src/main/java/se/gustavkarlsson/skylight/android/dagger/modules/CacheModule.java
package se.gustavkarlsson.skylight.android.dagger.modules;

import javax.inject.Singleton;

import dagger.Binds;
import dagger.Module;
import se.gustavkarlsson.skylight.android.cache.AuroraReportCache;
import se.gustavkarlsson.skylight.android.cache.DualLruAuroraReportCache;
import se.gustavkarlsson.skylight.android.cache.DualLruReportNotificationCache;
import se.gustavkarlsson.skylight.android.cache.ReportNotificationCache;

@Module
public abstract class CacheModule {

	@Binds
	@Singleton
	abstract AuroraReportCache bindAuroraReportCache(DualLruAuroraReportCache dualLruAuroraReportCache);

	@Binds
	@Singleton
	abstract ReportNotificationCache bindReportNotificationCache(DualLruReportNotificationCache dualLruReportNotificationCache);

}

File: app/src/main/kotlin/se/gustavkarlsson/skylight/android/dagger/modules/CacheModule.kt
package se.gustavkarlsson.skylight.android.dagger.modules

import android.content.Context
import dagger.Module
import dagger.Provides
import se.gustavkarlsson.skylight.android.entities.AuroraReport
import se.gustavkarlsson.skylight.android.services.SingletonCache
import se.gustavkarlsson.skylight.android.services_impl.cache.DualSingletonCache
import se.gustavkarlsson.skylight.android.services_impl.cache.auroraReportCacheSerializer
import javax.inject.Singleton

@Module
class CacheModule {

	@Provides
	@Singleton
	fun provideAuroraReportCache(context: Context): SingletonCache<AuroraReport> {
		return DualSingletonCache(
			CACHE_ID,
			AuroraReport.empty,
			auroraReportCacheSerializer,
			context
		)
	}

	companion object {
		private const val CACHE_ID = "latest-aurora-report"
	}
}
